File: iOS/小码哥/聊天界面/聊天界面/ViewController.m
//
//  ViewController.m
//  聊天界面
//
//  Created by carl on 2020/11/20.
//

#import "ViewController.h"
#import "Message.h"
#import "MeCell.h"
#import "OtherCell.h"
#import "MessageCell.h"
#import "MessageCell2.h"

@interface ViewController () <UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (strong, nonatomic) NSMutableArray* messages;



@end

@implementation ViewController

- (NSMutableArray *)messages {
    if (_messages == nil) {
        NSString *path = [[NSBundle mainBundle] pathForResource:@"messages" ofType:@"plist"];
        NSArray *dictArray = [NSArray arrayWithContentsOfFile:path];
        
        NSMutableArray *messagesArray = [NSMutableArray array];
        Message *lastMessage = nil;
        for (NSDictionary *dict in dictArray) {
            Message *message = [Message messageWithDict:dict];
            if (lastMessage == nil) {
                message.timeHidden = false;
            } else {
                message.timeHidden = [message.time isEqualToString:lastMessage.time];
            }
            [messagesArray addObject:message];
            lastMessage = message;
        }
        
        _messages = messagesArray;
    }
    return _messages;
}


- (void)viewDidLoad {
    [super viewDidLoad];
    
//    [self.tableView registerClass:[MeCell class] forCellReuseIdentifier:@"MeCell"];
//    [self.tableView registerClass:[OtherCell class] forCellReuseIdentifier:@"OtherCell"];
//    [self.tableView registerClass:[MessageCell class] forCellReuseIdentifier:@"MessageCell"];
}

#pragma mark - TABLEVIEW DATASOURCE AND DELEGATE
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.messages.count;
}

#pragma mark - Use seperate cells to show message
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    Message *message = self.messages[indexPath.row];
    
    MessageCell2 *cell = [tableView dequeueReusableCellWithIdentifier:@"MessageCell2" forIndexPath:indexPath];
    
    cell.message = message;
    
    return cell;
    
//    if (message.type == MessageTypeMe) {
//        MeCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MeCell" forIndexPath:indexPath];
//        cell.message = message;
//        return cell;
//    } else {
//        OtherCell *cell = [tableView dequeueReusableCellWithIdentifier:@"OtherCell" forIndexPath:indexPath];
//        cell.message = message;
//        return cell;
//    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    Message *message = self.messages[indexPath.row];

    return message.cellHeight;
}

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 200;
}



@end

File: iOS/小码哥/聊天界面Swift/聊天界面Swift/ViewController.swift
//
//  ViewController.swift
//  聊天界面Swift
//
//  Created by carl on 2020/11/21.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    @IBOutlet weak var tableView: UITableView!
    
    lazy var messages: [Message] = {
        let path = Bundle.main.path(forResource: "messages", ofType: "plist")
        let dictArray = NSArray(contentsOfFile: path!)
        let data = NSMutableArray()
        
        var lastMessage: Message?
        
        for dict in dictArray! {
            var message = Message(with: (dict as! [String: Any]))
            
            if lastMessage == nil {
                message.timeHidden = false
            } else {
                message.timeHidden = (message.time == lastMessage!.time) ? true : false
            }
            
            data.add(message)
        }
        
        return data as! [Message]
    } ()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.register(UINib(nibName: "MessageCell2", bundle: nil), forCellReuseIdentifier: "Cell")
        
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.messages.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath) as! MessageCell2
        
        let message = messages[indexPath.row]
        
        cell.message = message
        
        return cell
    }
    
//    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
//        return 350
//    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let message = messages[indexPath.row];

        return message.cellHeight ?? 0
    }

    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        return 100
    }
}

